drivers/sensors/pmw33xx_common.c

```diff
@@ -0,0 +1,208 @@
+    }
+    // tNCS-SCLK, 10ns
+    wait_us(1);
+    return true;
+}
+
+void pmw33xx_write(uint8_t sensor, uint8_t reg_addr, uint8_t data) {
+    if (!pmw33xx_spi_start(sensor)) {
+        return;
+    }
+
+    if (reg_addr != REG_Motion_Burst) {
+        in_burst[sensor] = false;
+    }
+
+    // send address of the register, with MSBit = 1 to indicate it's a write
+    uint8_t command[2] = {reg_addr | 0x80, data};
+    spi_transmit(command, sizeof(command));
+
+    // tSCLK-NCS for write operation is 35us
+    wait_us(35);
+    spi_stop();
+
+    // tSWW/tSWR (=18us) minus tSCLK-NCS. Could be shortened, but it looks like
+    // a safe lower bound
+    wait_us(145);
+}
+
+uint8_t pmw33xx_read(uint8_t sensor, uint8_t reg_addr) {
+    if (!pmw33xx_spi_start(sensor)) {
+        return 0;
+    }
+
+    // send adress of the register, with MSBit = 0 to indicate it's a read
+    spi_write(reg_addr & 0x7f);
+    // tSRAD (=160us)
+    wait_us(160);
+    uint8_t data = spi_read();
+
+    // tSCLK-NCS, 120ns
+    wait_us(1);
+    spi_stop();
+
+    //  tSRW/tSRR (=20us) mins tSCLK-NCS
+    wait_us(19);
+    return data;
+}
+
+bool pmw33xx_check_signature(uint8_t sensor) {
+    uint8_t signature_dump[3] = {
+        pmw33xx_read(sensor, REG_Product_ID),
+        pmw33xx_read(sensor, REG_Inverse_Product_ID),
+        pmw33xx_read(sensor, REG_SROM_ID),
+    };
+
+    return memcmp(pmw33xx_firmware_signature, signature_dump, sizeof(signature_dump)) == 0;
+}
+
+void pmw33xx_upload_firmware(uint8_t sensor) {
+    // Datasheet claims we need to disable REST mode first, but during startup
+    // it's already disabled and we're not turning it on ...
+    // pmw33xx_write(REG_Config2, 0x00);  // disable REST mode
+    pmw33xx_write(sensor, REG_SROM_Enable, 0x1d);
+    wait_ms(10);
+    pmw33xx_write(sensor, REG_SROM_Enable, 0x18);
+
+    if (!pmw33xx_spi_start(sensor)) {
+        return;
+    }
+
+    spi_write(REG_SROM_Load_Burst | 0x80);
+    wait_us(15);
+
+    for (size_t i = 0; i < PMW33XX_FIRMWARE_LENGTH; i++) {
+        spi_write(pgm_read_byte(pmw33xx_firmware_data + i));
+        wait_us(15);
+    }
+    wait_us(200);
+
+    pmw33xx_read(sensor, REG_SROM_ID);
+    pmw33xx_write(sensor, REG_Config2, 0x00);
+}
+
+bool pmw33xx_init(uint8_t sensor) {
+    if (sensor >= pmw33xx_number_of_sensors) {
+        return false;
+    }
+    spi_init();
+
+    // power up, need to first drive NCS high then low. the datasheet does not
+    // say for how long, 40us works well in practice.
+    if (!pmw33xx_spi_start(sensor)) {
+        return false;
+    }
+    wait_us(40);
+    spi_stop();
+    wait_us(40);
+
+    pmw33xx_write(sensor, REG_Power_Up_Reset, 0x5a);
+    wait_ms(50);
+
+    // read registers and discard
+    pmw33xx_read(sensor, REG_Motion);
+    pmw33xx_read(sensor, REG_Delta_X_L);
+    pmw33xx_read(sensor, REG_Delta_X_H);
+    pmw33xx_read(sensor, REG_Delta_Y_L);
+    pmw33xx_read(sensor, REG_Delta_Y_H);
+
+    pmw33xx_upload_firmware(sensor);
```

Comment:
this should return a bool, as it can fail and then we'd just blindly would continue. I guess it'll be harmless, it depends a bit on how "correct" you want this to be.

Reply:
Done.